src/goto-synthesizer/expr_enumerator.cpp

```diff
@@ -316,7 +316,18 @@ exprt binary_functional_enumeratort::instantiate(const expr_listt &exprs) const
     exprs.size() == 2,
     "number of arguments should be 2: " + integer2string(exprs.size()));
   if(op_id == ID_equal)
+  {
+    if(exprs.front().type() != exprs.back().type())
+      return true_exprt();
```

Comment:
Why is this the right thing to do?

Reply:
It is safe to do this. But it is not a perfect fix for the problem
The problem is that unsignedbv types with different width are actually different types. When build the grammar for the enumerator, I put all unsignedbv variables as terminals for one non-terminal. Then the grammar could enumerate equal expression between two variables with different types. I let it enumerate true here to avoid this happen.
Example
N -> N == N | N + N | `a` | `b` | `c`
where `a` and `b` are of `unsigned` type and `c` is of `unsigned long`.
I am thinking about building well-typed grammar. But maybe a better fix here is to allow euqal expression between unsignedbv types with different width, which is allowed in C. What do you think?

Reply:
If it's strictly about `unsigned(bv)`, then widening via typecast seems safe.